Dense vector docvalues: add base64 format

ah89 opened the pull request:
Add base64/binary docvalue output for dense_vector fetch/doc_value fields. Fields while keeping array as the default docvalue format. Add REST YAML coverage for base64/binary docvalue_fields

Closes #138189

ah89 committed:
Dense vector docvalues: add base64 format

Add base64/binary docvalue output for dense_vector fetch/doc_value
fields. Fields while keeping array as the default docvalue format.
Add REST YAML coverage for base64/binary docvalue_fields

Closes #138189

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:
```diff
@@ -2603,8 +2603,15 @@ protected Object parseSourceValue(Object value) {
+            if (format == null || "array".equals(format)) {
+                return DocValueFormat.DENSE_VECTOR;
+            }
+            if ("base64".equals(format) || "binary".equals(format)) {
```
I think we just want to have a single name either "base64" or "binary", not both.
I would vote for "base64" as it is more explicit.

@benwtrent what do you think?

ah89 replied:
As @benwtrent [suggested](https://github.com/elastic/elasticsearch/pull/140094#discussion_r2661817680), we kept `binary`.

mayya-sharipova commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/205_dense_vector_docvalue_fields_format.yml:
```diff
@@ -0,0 +1,107 @@
+        - method: POST
+          path: /_search
+          capabilities: [ dense_vector_docvalue_fields ]
+      cluster_features: [ "mapper.vectors.generic_vector_format" ]
```
You need to define a new cluster_feature in [MapperFeatures](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/mapper/MapperFeatures.java#L62): e.g `BASE64_VECTOR_FORMAT`

ah89 replied:
Added `BASE64_VECTOR_FORMAT` to MapperFeatures.java

mayya-sharipova commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/205_dense_vector_docvalue_fields_format.yml:
```diff
@@ -0,0 +1,107 @@
+      capabilities:
+        - method: POST
+          path: /_search
+          capabilities: [ dense_vector_docvalue_fields ]
```
We need to introduce a new capability in `SearchCapability` class.

ah89 replied:
Added `DENSE_VECTOR_DOCVALUE_FIELDS_FORMAT` to SearchCapabilities.java

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/VectorDVLeafFieldData.java:
```diff
@@ -210,6 +220,108 @@ public FormattedDocValues getFormattedValues(DocValueFormat format) {
+        }
+    }
+
+    private class Base64FloatDocValues implements FormattedDocValues {
```
Looks like  `Base64FloatDocValues` shares most of the code with FloatDocValues.

Can we instead have just a single class `FloatDocValues` with an extra instance variable "format" that can be used in  `nextValue` method to define what format output the value?

ah89 committed:
[#138189] Fix dense_vector docvalue formats

- REST: remove unused name field from
  205_dense_vector_docvalue_fields_format.yml
- Mapper: drop binary format; support only base64 and array
- Mapper: map base64 to DocValueFormat.BINARY
- Mapper: map array to DocValueFormat.DENSE_VECTOR
- REST: add coverage for format: array
- Features: add BASE64_VECTOR_FORMAT to MapperFeatures
- Caps: add DENSE_VECTOR_DOCVALUE_FIELDS_FORMAT to SearchCapabilities
- Tests: extend DenseVectorFieldTypeTests for bit and bfloat16; assert
  binary is rejected
- Refactor: remove unused bits param from Base64ByteDocValues
- Refactor: merge Base64FloatDocValues into FloatDocValues via base64
  flag
- Perf: reuse byte[] and ByteBuffer for base64 encoding
- Fix: decode bfloat16 docvalues with decodeBFloat16DenseVector

Verified by running DenseVectorFieldTypeTests.

elasticsearchmachine commented:
Pinging @elastic/es-search-relevance (Team:Search Relevance)

ah89 commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/VectorDVLeafFieldData.java:
```diff
@@ -210,6 +220,108 @@ public FormattedDocValues getFormattedValues(DocValueFormat format) {
+
+        Base64ByteDocValues(int dims, boolean bits) {
+            this.dims = dims;
+            this.bits = bits;
```
Modified to remove the unused `bits` parameter from Base64ByteDocValues.

ah89 commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/VectorDVLeafFieldData.java:
```diff
@@ -210,6 +220,108 @@ public FormattedDocValues getFormattedValues(DocValueFormat format) {
+                    return false;
+                }
+                BytesRef ref = binary.binaryValue();
+                VectorEncoderDecoder.decodeDenseVector(indexVersion, ref, vector);
```
Updated `BFloat16DocValues` to correctly use `VectorEncoderDecoder.decodeBFloat16DenseVector` for decoding.

benwtrent commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:
```diff
@@ -2603,8 +2603,15 @@ protected Object parseSourceValue(Object value) {
+            if (format == null || "array".equals(format)) {
+                return DocValueFormat.DENSE_VECTOR;
+            }
+            if ("base64".equals(format)) {
```
we should call this `binary`

ah89 replied:
Done. This commit explicitly replaces “base64” with “binary” as the supported encoded output name, and updates tests/error messages accordingly.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:
```diff
@@ -2603,8 +2603,15 @@ protected Object parseSourceValue(Object value) {
+            }
+            throw new IllegalArgumentException(
+                "Field [" + name() + "] of type [" + typeName() + "] doesn't support format [" + format + "]."
+            );
```
Let's all add what are supported formats to this error message

mayya-sharipova commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/VectorDVLeafFieldData.java:
```diff
@@ -202,14 +222,73 @@ void decodeDenseVector(IndexVersion indexVersion, BytesRef vectorBR, float[] vec
         };
     }
 
+    private class Base64ByteDocValues implements FormattedDocValues {
```
Similarly with FloatDocValues, can we combine Base64ByteDocValues and ByteDocValues into a single class as they mostly share the same code?

mayya-sharipova commented:
@ah89 Thanks for iterating on this! I've left some additional comments, but it is very close to be done to my mind.

ah89 committed:
Merge branch 'main' into feature/vector-docvalues-base64-format

ah89 committed:
Refactor dense_vector docvalues format handling

Replace base64 with binary encoded output.

Unify and simplify dense_vector docvalues format selection by
replacing base64 with binary as the supported encoded
output. Update related tests and error messages for clarity
and maintainability, and remove redundant or obsolete test
cases. Improve flexibility and consistency in dense vector
field output formats.

Relates to #138189

ah89 commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/VectorDVLeafFieldData.java:
```diff
@@ -189,11 +200,20 @@ public int docValueCount() {
         @Override
         public Object nextValue() {
+            if (base64) {
+                scratchBuffer.clear();
```
Done. The refactor cleanup removes redundant buffer state operations and redundant fields.

ah89 committed:
Remove legacy base64 vector format support

Deprecate base64 vector format in favor of generic and binary.

Deprecate the base64 vector format for dense vector docvalue
fields in favor of generic and binary formats.

Aligns naming with current conventions and simplifies feature
management by eliminating legacy terminology.

Relates to #138189

ah89 commented:
[Done](https://github.com/elastic/elasticsearch/pull/140094#issuecomment-3709465818). The Feature flag removed from [MapperFeatures](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/mapper/MapperFeatures.java)

mayya-sharipova reviewed (approved):
@ah89 Thanks for iterating, nice progress.

I've a small comment about error message. Please address it, but otherwise LGTM

benwtrent commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/205_dense_vector_docvalue_fields_format.yml:
```diff
@@ -0,0 +1,179 @@
+          vec_bfloat16_indexed: [ 1.5, 2.0, -3.25 ]
+          vec_bfloat16_dv: [ 1.5, 2.0, -3.25 ]
+  - do:
+      indices.refresh: { }
```
can we add a second doc where we index the values as binary directly? This we we test array vs. binary for vectors that are arrays and binary on index :)

benwtrent commented on server/src/main/java/org/elasticsearch/index/mapper/vectors/DenseVectorFieldMapper.java:
```diff
@@ -2603,8 +2603,20 @@ protected Object parseSourceValue(Object value) {
+            return switch (format) {
+                case null -> DocValueFormat.DENSE_VECTOR;
+                case "array" -> DocValueFormat.DENSE_VECTOR;
+                case "binary" -> DocValueFormat.BINARY;
```
modern java is getting so much nicer :D!

benwtrent reviewed (approved):
I think this is good. I do think we need one more set of test cases to ensure the `binary -> binary` retrieval is good :)

Approving assuming you will add those tests before merging.

ah89 committed:
Merge branch 'main' into feature/vector-docvalues-base64-format

ah89 committed:
Adds tests for dense_vector binary docvalue_fields

This commit adds a test case to REST test that indexes vectors using
binary encoded strings. It verifies that these documents can be
retrieved correctly via docvalue_fields in both array and binary
output formats. This ensures that the dense vector format handling is
robust to the ingestion format.

Also clarifies supported formats by removing null from the list of
accepted docvalue_field formats in error messages.

Closes #138189

ah89 committed:
Merge branch 'main' into feature/vector-docvalues-base64-format

ah89 committed:
Merge branch 'main' into feature/vector-docvalues-base64-format
